pkg/cloudlogging/puller_test.go

```diff
@@ -98,7 +98,109 @@ func TestPull(t *testing.T) {
 				t.Errorf("Process(%+v) got result diff (-want, +got): %v", tc.name, diff)
 			}
 
-			if diff := cmp.Diff(tc.wantReq, tc.server.req, protocmp.Transform()); diff != "" {
+			if diff := cmp.Diff(tc.wantReq, tc.server.listReq, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got request diff (-want, +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+func TestStreamPull(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		server        *fakeServer
+		filter        string
+		wantReq       *loggingpb.TailLogEntriesRequest
+		wantResult    []*loggingpb.LogEntry
+		wantNum       int
+		wantErrSubstr string
+	}{
+		{
+			name:   "success",
+			filter: "test-filter",
+			server: &fakeServer{
+				tailResp: &loggingpb.TailLogEntriesResponse{
+					Entries: []*loggingpb.LogEntry{{LogName: "test1"}, {LogName: "test2"}},
+				},
+			},
+			wantResult: []*loggingpb.LogEntry{{LogName: "test1"}, {LogName: "test2"}},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantNum: 2,
+		},
+		{
+			name:   "failed_to_pull",
+			filter: "test-filter",
+			server: &fakeServer{
+				injectedErr: fmt.Errorf("injected error"),
+			},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantErrSubstr: "injected error",
+			wantNum:       1,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ch := make(chan *loggingpb.LogEntry)
+			var gotLogs []*loggingpb.LogEntry
+			errCh := make(chan error)
+			done := make(chan struct{}, 1)
+			t.Cleanup(func() {
+				close(done)
+			})
+
+			ctx := context.Background()
+			go func() {
+				for l := range ch {
+					gotLogs = append(gotLogs, l)
+					if len(gotLogs) == tc.wantNum {
+						done <- struct{}{}
+						break
+					}
+				}
+			}()
+
+			fakeClient, streamClient := setupFakeStreamClient(t, ctx, tc.server)
+			p := NewPuller(
+				ctx,
+				fakeClient,
+				testResource,
+				WithRetry(retry.WithMaxRetries(0, retry.NewFibonacci(500*time.Millisecond))),
+			)
+
+			go p.StreamPull(ctx, tc.filter, ch, errCh, streamClient)
+
+			select {
+			case <-done:
+				err := streamClient.CloseSend()
+				if err != nil {
+					t.Fatalf("failed to close streamClient %v", err)
+				}
```

Comment:
We should cancel context? That's what will happen if a user hit ctrl+c.

Reply:
Use `context.WithCancel` now.